Sources/GRPC/Docs.docc/Proposals/0001-stub-api.md

```diff
@@ -0,0 +1,1100 @@
+  // message which would make the common case straighforward:
+  let message = try await echo.get(request: request)
+  print("'Get' received '\(message.text)'")
+}
+
+func clientStreaming(echo: some EchoClientProtocol) async throws {
+  // Make the request:
+  let request = ClientRequest.Stream<Echo.Request> { writer in
+    for text in ["foo", "bar", "baz"] {
+      try await writer.write(Echo.Request(text: text))
+    }
+  }
+
+  // Execute the RPC:
+  try await echo.collect(request: request) { response in
+    // (Same as for the unary "Get")
+  }
+}
+
+func serverStreaming(echo: some EchoClientProtocol) async throws {
+  // Make the request, adding metadata:
+  let request = ClientRequest.Single(
+    message: Echo.Request(text: "foo bar baz"),
+    metadata: ["foo": "bar"],
+  )
+
+  // Execute the RPC (most verbose API):
+  try await echo.expand(request: request) { response in
+    switch response.result {
+    case .success(let body):
+      print("'Expand' accepted with metadata '\(body.metadata)'")
+      do {
+        for try await part in body.bodyParts {
+          switch part {
+          case .message(let message):
+            print("'Expand' received message '\(message.text)'")
+          case .trailers(let metadata):
+            print("'Expand' received trailers '\(metadata)'")
+          }
+        }
+      } catch let error as RPCError {
+        print("'Expand' failed with error '\(error.code)' and metadata '\(error.metadata)'")
+      } catch {
+        print("'Expand' failed with error '\(error)'")
+      }
+    case .failure(let error):
+      print("'Expand' rejected with error '\(error.code)' and metadata '\(error.metadata)'")
+    }
+  }
+
+  // Execute the RPC (sugared API):
+  await echo.expand(request: request) { response in
+    print("'Expand' received metadata '\(response.metadata)'")
+    do {
+      for try await message in response.messages {
+        print("'Expand' received '\(message.text)'")
+      }
+    } catch let error as RPCError {
+      print("'Expand' failed with error '\(error.code)' and metadata '\(error.metadata)'")
+    } catch {
+      print("'Expand' failed with error '\(error)'")
+    }
+  }
+
+  // Note: there is no generated 'default' handler, the function body defines
+  // the lifetime of the RPC and the RPC is cancelled once the closure exits.
+  // Therefore escaping the message sequence would result in the sequence
+  // throwing an error. It must be consumed from within the handler.
+}
+
+func bidirectional(echo: some EchoClient) async throws {
+  // Make the request, adding metadata:
+  let request = ClientRequest.Stream<Echo.Response>(metadata: ["foo": "bar"]) { writer in
+    for text in ["foo", "bar", "baz"] {
+      try await writer.write(Echo.Request(text: text))
+    }
+  }
+
+  // Execute the RPC:
+  try await echo.collect(request: request) { response in
+    // (Same as for the server streaming "Expand")
+  }
```

Comment:
I find this spelling very interesting and a bit surprising. The problem that I see is that there are now two separate scopes - one for the request writer and one for the response. If you want to implement business logic where you write a certain request based on a response you received this becomes very hard and needs a layer of indirection.

From previous experience, I found it helpful to have the bidirectional streaming closure provide both the `responseStream` and the `requestWriter`.

Just some pseudo code for what I mean

```
try await someClient.bidistream(metadata: ...) { requestWriter, responseStream in
    try await requestWriter.write(...)
    
    let firstResponse = try await responseStream.first
    
    if firstResponse.someProperty {
        // different message written
    } else {
        // different message written
    }
}
```

Reply:
I see this a _good_ thing: if your input stream depends on the output stream then you should make the dependency explicit via an additional `AsyncStream` or similar. Splitting them up makes it harder to accidentally deadlock.

It does make some use cases harder, but they are still possible.

Reply:
IMO, this is really important. I have myself written a bunch of tests where driving both the requests and responses in tandem is important. This is mostly the case when you implement something like a distributed state machine over a bidirectional streaming RPC.

Having to go through an intermediate stream can work but has two downsides:
1. It is currently not implementable since we don't have a backpressure `AsyncSequence` that we could use here.
2. It becomes less performant since we now have another layer of indirection.

Reply:
Beyond the issues mentioned above, the request and response being tied together has an impact on the interceptor API: it effectively takes the same shape as the function which isn't at all ergonomic to use. The knock on from this is that the client API diverges even further from the server API.

I'm confident that this is the correct spelling for the base API: it separates the input from the output and makes it hard for users to misuse the API. I

> It is currently not implementable since we don't have a backpressure AsyncSequence that we could use here.

We also don't have this API right now! There are backpressure `AsyncSequence`s which do work: `AsyncChannel` (sure, a bit slow as the buffer size is 1, but it works and is probably suitable for most uses cases)  and NIOs producer. There are others in the works too.

> It becomes less performant since we now have another layer of indirection.

True, but if your two streams depend on each other then the RTT will likely dominate anyway. 

>  I have myself written a bunch of tests where driving both the requests and responses in tandem is important.

If requested enough we can add a utility or sugar for this. I don't think it should be the base API though.